internal/controller/management/serviceaccounts/serviceaccounts.go

```diff
@@ -0,0 +1,239 @@
+func SetupReconcilerWithManager(kargoMgr manager.Manager, cfg ReconcilerConfig) error {
+	return ctrl.NewControllerManagedBy(kargoMgr).
+		For(&corev1.ServiceAccount{}).
+		WithEventFilter(
+			predicate.Funcs{
+				DeleteFunc: func(event.DeleteEvent) bool {
+					// We're only interested in soft deletes
+					return false
+				},
+			},
+		).
+		Complete(newReconciler(kargoMgr.GetClient(), cfg))
+}
+
+func newReconciler(kubeClient client.Client, cfg ReconcilerConfig) *reconciler {
+	return &reconciler{
+		cfg:    cfg,
+		client: kubeClient,
+	}
+}
+
+// Reconcile is part of the main Kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"serviceAccount", req.NamespacedName.Name,
+		"serviceAccount.namespace", req.NamespacedName.Namespace,
+	)
+	ctx = logging.ContextWithLogger(ctx, logger)
+	logger.Debug("reconciling ServiceAccount")
+
+	sa := &corev1.ServiceAccount{}
+	if err := r.client.Get(ctx, req.NamespacedName, sa); err != nil {
+		if kubeerr.IsNotFound(err) {
+			// Ignore if not found. This can happen if the ServiceAccount was deleted
+			// after the current reconciliation request was issued.
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf(
+			"error getting ServiceAccount %q in namespace %q: %w",
+			req.NamespacedName.Name, req.NamespacedName.Namespace, err,
+		)
+	}
+
+	if (sa.DeletionTimestamp != nil || !r.hasControllerLabel(sa)) &&
+		controllerutil.ContainsFinalizer(sa, kargoapi.FinalizerName) {
+		// Ensure non-existence of RoleBindings that grant this controller
+		// ServiceAccount access to read Secrets in all Project namespaces.
+		if err := r.removeControllerPermissions(ctx, req.NamespacedName); err != nil {
+			return ctrl.Result{}, err
+		}
+		if controllerutil.RemoveFinalizer(sa, kargoapi.FinalizerName) {
+			if err := r.client.Update(ctx, sa); err != nil {
+				return ctrl.Result{}, fmt.Errorf(
+					"error removing finalizer from ServiceAccount %q in namespace %q: %w",
+					sa.Name, sa.Namespace, err,
+				)
+			}
+			logger.Debug("removed finalizer from ServiceAccount")
+		}
+		logger.Debug("done reconciling ServiceAccount")
+		return ctrl.Result{}, nil
+	}
+
+	if sa.DeletionTimestamp == nil && r.hasControllerLabel(sa) {
+		// Ensure the existence of RoleBindings that grant this controller
+		// ServiceAccount access to read Secrets in all Project namespaces.
+		if controllerutil.AddFinalizer(sa, kargoapi.FinalizerName) {
+			if err := r.client.Update(ctx, sa); err != nil {
+				return ctrl.Result{}, fmt.Errorf(
+					"error adding finalizer to ServiceAccount %q in namespace %q: %w",
+					sa.Name, sa.Namespace, err,
+				)
+			}
+			logger.Debug("added finalizer to ServiceAccount")
+			if err := r.ensureControllerPermissions(ctx, req.NamespacedName); err != nil {
```

Comment:
Do we only need to run this if `controllerutil.AddFinalizer(sa, kargoapi.FinalizerName) == true`?

Reply:
You caught a 🐞

Reply:
Fixed in a837a61b9a1961cdb6b27d584170b6f2b333a1a7